agent/xds/server.go

```diff
@@ -0,0 +1,418 @@
+			//     dropping or misrouting traffic due to inconsitencies. This is the
+			//     main win of ADS after all - we get to control this order.
+			//  2. Non-determinsic order of complex protobug responses which are
+			//     compared for non-exact JSON equivalence makes the tests uber-messy
+			//     to handle
+			for _, typeURL := range []string{ClusterType, EndpointType, RouteType, ListenerType} {
+				handler := handlers[typeURL]
+				err := handler.SendIfNew(cfgSnap, configVersion, &nonce)
+				if err != nil {
+					return err
+				}
+			}
+		}
+	}
+}
+
+type xDSType struct {
+	typeURL   string
+	stream    ADSStream
+	req       *v2.DiscoveryRequest
+	lastNonce string
+	// lastVersion is the version that was last sent to the proxy. It is needed
+	// because we don't want to send the same version more than once.
+	// req.VersionInfo may be an older version than the most recent once sent in
+	// two cases: 1) if the ACK wasn't received yet and `req` still points to the
+	// previous request we already responded to and 2) if the proxy rejected the
+	// last version we sent with a Nack then req.VersionInfo will be the older
+	// version it's hanging on to.
+	lastVersion uint64
+	resources   func(cfgSnap *proxycfg.ConfigSnapshot, token string) ([]proto.Message, error)
+}
+
+func (t *xDSType) Recv(req *v2.DiscoveryRequest) {
+	if t.lastNonce == "" || t.lastNonce == req.GetResponseNonce() {
+		t.req = req
+	}
+}
+
+func (t *xDSType) SendIfNew(cfgSnap *proxycfg.ConfigSnapshot, version uint64, nonce *uint64) error {
+	if t.req == nil {
+		return nil
+	}
+	if t.lastVersion >= version {
+		// Already sent this version
+		return nil
+	}
+	resources, err := t.resources(cfgSnap, tokenFromStream(t.stream))
+	if err != nil {
+		return err
+	}
+	if resources == nil || len(resources) == 0 {
+		// Nothing to send yet
+		return nil
+	}
+
+	// Note we only increment nonce when we actually send - not important for
+	// correctness but makes tests much simpler when we skip a type like Routes
+	// with nothing to send.
+	*nonce++
+	nonceStr := fmt.Sprintf("%08x", *nonce)
+	versionStr := fmt.Sprintf("%08x", version)
+
+	resp, err := createResponse(t.typeURL, versionStr, nonceStr, resources)
+	if err != nil {
+		return err
+	}
+
+	err = t.stream.Send(resp)
+	if err != nil {
+		return err
+	}
+	t.lastVersion = version
+	t.lastNonce = nonceStr
+	return nil
+}
+
+func tokenFromStream(stream ADSStream) string {
+	return tokenFromContext(stream.Context())
+}
+
+func tokenFromContext(ctx context.Context) string {
+	md, ok := metadata.FromIncomingContext(ctx)
+	if !ok {
+		return ""
+	}
+	toks, ok := md["x-consul-token"]
+	if ok && len(toks) > 0 {
+		return toks[0]
+	}
+	return ""
+}
+
+// IncrementalAggregatedResources implements discovery.AggregatedDiscoveryServiceServer
+func (s *Server) IncrementalAggregatedResources(_ discovery.AggregatedDiscoveryService_IncrementalAggregatedResourcesServer) error {
+	return errors.New("not implemented")
```

Comment:
What are the implications of implementing-but-not-implementing this method?

Reply:
It's an optional API that Envoy has to be configured to use currently. We should totally implement it in the future as it's more scalable that reloading the entire state every time one thing changes but not necessary as it's an optional protocol extension.